treegp/kernels.py

```diff
@@ -0,0 +1,392 @@
+                     2-element iterable, which will be taken to be the min and max value for every
+                     theta element, or an [ntheta, 2] array indicating bounds on each of ntheta
+                     elements.
+    """
+    def __init__(self, invLam=None, scale_length=None, bounds=(-5,5)):
+        if scale_length is not None:
+            if invLam is not None:
+                raise TypeError("Cannot set both invLam and scale_length in AnisotropicRBF.")
+            invLam = np.diag(1./np.array(scale_length)**2)
+
+        self.ndim = invLam.shape[0]
+        self.ntheta = self.ndim*(self.ndim+1)//2
+        self._d = np.diag_indices(self.ndim)
+        self._t = np.tril_indices(self.ndim, -1)
+        self.set_params(invLam)
+        bounds = np.array(bounds)
+        if bounds.ndim == 1:
+            bounds = np.repeat(bounds[None, :], self.ntheta, axis=0)
+        assert bounds.shape == (self.ntheta, 2)
+        self._bounds = bounds
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        from scipy.spatial.distance import pdist, cdist, squareform
+        X = np.atleast_2d(X)
+
+        if Y is None:
+            dists = pdist(X, metric='mahalanobis', VI=self.invLam)
+            K = np.exp(-0.5 * dists**2)
+            K = squareform(K)
+            np.fill_diagonal(K, 1)
+        else:
+            if eval_gradient:
+                raise ValueError(
+                    "Gradient can only be evaluated when Y is None.")
+            dists = cdist(X, Y, metric='mahalanobis', VI=self.invLam)
+            K = np.exp(-0.5 * dists**2)
+
+        if eval_gradient:
+            if self.hyperparameter_cholesky_factor.fixed:
+                return K, np.empty((X.shape[0], X.shape[0], 0))
+            else:
+                # dK_pq/dth_k = -0.5 * K_pq *
+                #               ((x_p_i-x_q_i) * dInvLam_ij/dth_k * (x_q_j - x_q_j))
+                # dInvLam_ij/dth_k = dL_ij/dth_k * L_ij.T  +  L_ij * dL_ij.T/dth_k
+                # dL_ij/dth_k is a matrix with all zeros except for one element.  That element is
+                # L_ij if k indicates one of the theta parameters landing on the Cholesky diagonal,
+                # and is 1.0 if k indicates one of the thetas in the lower triangular region.
+                L_grad = np.zeros((self.ntheta, self.ndim, self.ndim), dtype=float)
+                L_grad[(np.arange(self.ndim),)+self._d] = self._L[self._d]
+                L_grad[(np.arange(self.ndim, self.ntheta),)+self._t] = 1.0
+
+                half_invLam_grad = np.dot(L_grad, self._L.T)
+                invLam_grad = half_invLam_grad + np.transpose(half_invLam_grad, (0, 2, 1))
+
+                dX = X[:, np.newaxis, :] - X[np.newaxis, :, :]
+                dist_grad = np.einsum("ijk,lkm,ijm->ijl", dX, invLam_grad, dX)
+                K_gradient = -0.5 * K[:, :, np.newaxis] * dist_grad
+                return K, K_gradient
```

Comment:
A lot of this gets easier if we use JAX and autodifferentiation (e.g., as done in https://tinygp.readthedocs.io/en/latest/). Using `tinygp` (and/or JAX fundamentals) could be a possible long-term revision to help scale this code up for larger data (as JAX can better leverage resources like GPUs, etc.).

Reply:
Indeed, but I guess this part is used for the Maximum Likelihood part and not when using `treecorr`. But it could be useful indeed.